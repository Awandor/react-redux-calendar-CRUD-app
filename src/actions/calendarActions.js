import { fetchWithToken } from "../helpers/fetchHelper";
import { types } from "../types/types";
import Swal from "sweetalert2";
import { adjustEventsDates } from "../helpers/adjustEventsDates";


// ========================================
// Active event
// ========================================

export const eventSetActiveStoreAction = ( event ) => ( {
  type: types.eventSetActive,
  payload: event
} );

export const eventClearActiveStoreAction = () => ( { type: types.eventClearActive } );


// ========================================
// Create new event
// ========================================

export const eventStartAddNewEventAction = ( event ) => {

  console.log( 'evento recibido por eventStartAddNewEventAction', event );

  // Como es una Acción asíncrona retornamos una función con async

  return async( dispatch, getState ) => {

    try {

      const resp = await fetchWithToken( 'events', event, 'POST' );

      const body = await resp.json(); // await porque json() retorna una promesa

      // console.log( 'eventStartAddNewEventAction', body );

      if ( body.ok ) {

        // Vamos a leer del Store el uid y el name con getState

        const { uid, name } = getState().auth;

        const newEvent = {
          ...event,
          id: body.event.id,
          user: {
            _id: uid,
            name
          }
        };

        // console.log( 'newEvent', newEvent );

        dispatch( eventAddNewEventStoreAction( newEvent ) );

      } else {

        Swal.fire( 'Error', body.msg, 'error' );

      }

    } catch ( error ) {

      Swal.fire( 'Error', error, 'error' );

    }

  };

};

export const eventAddNewEventStoreAction = ( event ) => ( {
  type: types.eventAddNew,
  payload: event
} );


// ========================================
// Load all events 
// ========================================

export const eventStartLoadAllEventsAction = () => {

  // Como es una Acción asíncrona retornamos una función con async

  return async( dispatch, getState ) => {

    try {

      const resp = await fetchWithToken( 'events' ); // es un GET

      const body = await resp.json(); // await porque json() retorna una promesa

      // console.log( 'eventStartLoadAllEventsAction', body );

      if ( body.ok ) {

        const events = adjustEventsDates( body.events );

        dispatch( eventLoadAllEventsAction( events ) );

      } else {

        Swal.fire( 'Error', body.msg, 'error' );

      }

    } catch ( error ) {

      Swal.fire( 'Error', error, 'error' );

    }

  };

};

export const eventLoadAllEventsAction = ( events ) => ( { type: types.eventAllLoaded, payload: events } );


// ========================================
// Update event
// ========================================

export const eventStartUpdateEventAction = ( event ) => {

  // Como es una Acción asíncrona retornamos una función con async

  return async( dispatch, getState ) => {

    try {

      const resp = await fetchWithToken( `events/${event.id}`, event, 'PUT' );

      const body = await resp.json(); // await porque json() retorna una promesa

      // console.log( 'eventStartUpdateEventAction', body );

      if ( body.ok ) {

        dispatch( eventUpdateStoreAction( event ) );

      } else {

        Swal.fire( 'Error', body.msg, 'error' );

      }

    } catch ( error ) {

      Swal.fire( 'Error', error, 'error' );

    }

  };

};

export const eventUpdateStoreAction = ( event ) => ( {
  type: types.eventUpdate,
  payload: event
} );


// ========================================
// Delete event
// ========================================

export const eventStartDeleteAction = ( event ) => {

  // Como es una Acción asíncrona retornamos una función con async

  return async( dispatch ) => {

    try {

      const resp = await fetchWithToken( `events/${event.id}`, {}, 'DELETE' );

      const body = await resp.json(); // await porque json() retorna una promesa

      console.log( 'eventStartDeleteAction', body );

      if ( body.ok ) {

        dispatch( eventDeleteStoreAction( event ) );

      } else {

        Swal.fire( 'Error', body.msg, 'error' );

      }

    } catch ( error ) {

      Swal.fire( 'Error', error, 'error' );

    }

  };

};

export const eventDeleteStoreAction = ( event ) => ( {
  type: types.eventDelete,
  payload: event
} );


// ========================================
// Clear Calendar
// ========================================

export const eventClearCalendarStoreAction = ( event ) => ( {
  type: types.eventClearCalendarStore
} );
