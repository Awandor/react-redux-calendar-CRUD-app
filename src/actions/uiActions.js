import { types } from "../types/types";


export const uiOpenModalStoreAction = () => ( {
  type: types.uiOpenModalAction
} );

export const uiCloseModalStoreAction = () => ( {
  type: types.uiCloseModalAction
} );
