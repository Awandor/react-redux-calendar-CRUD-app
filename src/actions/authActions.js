import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetchHelper";
import { types } from "../types/types";
import { eventClearCalendarStoreAction } from "./calendarActions";

// ========================================
// Login
// ========================================

export const authStartLoginAction = ( email, password ) => {

  // Como es una Acción asíncrona retornamos una función con async

  return async( dispatch ) => {

    // console.log( email, password );

    const resp = await fetchWithoutToken( 'auth/login', { email, password }, 'POST' );

    const body = await resp.json(); // await porque json() retorna una promesa

    // console.log( body );

    if ( body.ok ) {

      localStorage.setItem( 'token', body.token );

      localStorage.setItem( 'token-init-date', new Date().getTime() );

      dispatch( authLoginStoreAction( { uid: body.uid, name: body.name } ) );

    } else {

      Swal.fire( 'Error', body.msg, 'error' );

    }

  };

};

export const authLoginStoreAction = ( user ) => ( { type: types.authLogin, payload: user } );


// ========================================
// Register user
// ========================================

export const authStartRegisterAction = ( name, email, password ) => {

  // Como es una Acción asíncrona retornamos una función con async

  return async( dispatch ) => {

    // console.log( name, email, password );

    const resp = await fetchWithoutToken( 'auth/register', { name, email, password }, 'POST' );

    const body = await resp.json(); // await porque json() retorna una promesa

    // console.log( 'authStartRegisterAction', body );

    if ( body.ok ) {

      localStorage.setItem( 'token', body.token );

      localStorage.setItem( 'token-init-date', new Date().getTime() );

      dispatch( authLoginStoreAction( { uid: body.uid, name: body.name } ) );

    } else {

      Swal.fire( 'Error', body.msg, 'error' );

    }

  };

};


// ========================================
// Check token
// ========================================

export const authStartCheckingTokenAction = () => {

  return async( dispatch ) => {

    // La ruta renew del backend sólo necesita el token, no tiene body

    const resp = await fetchWithToken( 'auth/renew' );

    const body = await resp.json(); // await porque json() retorna una promesa

    // console.log( 'authStartCheckingTokenAction', body );

    if ( body.ok ) {

      localStorage.setItem( 'token', body.token );

      localStorage.setItem( 'token-init-date', new Date().getTime() );

      dispatch( authLoginStoreAction( { uid: body.uid, name: body.name } ) );

    } else {

      Swal.fire( 'Error', body.msg, 'error' );

      dispatch( authCheckingTokenFinishStoreAction() );

    }

  };

};

const authCheckingTokenFinishStoreAction = () => ( { type: types.authCheckingFinish } );


// ========================================
// Logout
// ========================================

export const startLogoutAction = () => {

  // La manejamos como si fuera una función asíncrona para tener acceso a dispatch

  return ( dispatch ) => {

    localStorage.removeItem( 'token' );
    localStorage.removeItem( 'token-init-date' );

    dispatch( logoutStoreAction() );

    dispatch( eventClearCalendarStoreAction() );

  };

};

export const logoutStoreAction = () => ( { type: types.authLogout } );
