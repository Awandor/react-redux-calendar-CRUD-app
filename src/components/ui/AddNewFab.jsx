import { useDispatch } from "react-redux";
import { uiOpenModalStoreAction } from "../../actions/uiActions";

export const AddNewFab = () => {

    // Necesitamos useDispatch para depachar Acciones

    const dispatch = useDispatch();

    const handleClickNewEvent = () => {

        console.log( 'Add' );

        dispatch( uiOpenModalStoreAction() );

    };

    return (
        <button className="btn btn-primary fab" onClick={handleClickNewEvent}><i className="fas fa-plus"></i></button>
    );

};