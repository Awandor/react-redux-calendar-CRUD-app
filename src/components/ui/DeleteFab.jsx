import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { eventStartDeleteAction } from '../../actions/calendarActions';

export const DeleteFab = () => {

    // Vamos a estar pendientes de cambios en activeEvent

    const { activeEvent } = useSelector( state => state.calendar );

    // Necesitamos useDispatch para despachar Acciones

    const dispatch = useDispatch();

    const handleClickDeleteEvent = () => {

        dispatch( eventStartDeleteAction( activeEvent ) );

    };

    return (
        <button className="btn btn-danger fab-delete" onClick={handleClickDeleteEvent}>
            <i className="fas fa-trash"></i>
        </button>
    );

};
