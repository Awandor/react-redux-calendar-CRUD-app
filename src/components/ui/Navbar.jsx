import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogoutAction } from '../../actions/authActions';

export const Navbar = () => {

    // Vamos a leer del Store la propiedad name

    const { name } = useSelector( state => state.auth );

    const dispatch = useDispatch();

    const handleLogout = () => {

        dispatch( startLogoutAction() );

    };

    return (
        <div className="navbar navbar-dark bg-dark mb-4">

            <span className="navbar-brand">{name}</span>

            <button className="btn btn-outline-danger" onClick={handleLogout}><i className="fas fa-sign-out-alt"></i> Salir</button>

        </div>
    );

};
