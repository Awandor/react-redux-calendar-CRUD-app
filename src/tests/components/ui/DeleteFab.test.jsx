import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom'; // Se importa sólo para tener ayuda en el tipado de jest

import { DeleteFab } from "../../../components/ui/DeleteFab";
import { eventStartDeleteAction } from "../../../actions/calendarActions";

// Creamos el mock del Store

const middlewares = [ thunk ];

const mockStore = configureStore( middlewares );

// Definimos el estado inicial

const initialState = { calendar: { activeEvent: {} } };

let store = mockStore( initialState );

// Hacemos un mock de dispatch, no vamos a evaluar la Acción, sólo que se realiza el dispatch de eventStartDeleteAction

store.dispatch = jest.fn();

// Hacemos un mock de eventStartDeleteAction

jest.mock( '../../../actions/calendarActions', ()=>( { eventStartDeleteAction: jest.fn() } ) );

const wrapper = mount(
    <Provider store={store}>
        <DeleteFab />
    </Provider>
);


describe( 'Pruebas en <DeleteFab />', () => {

    test( 'debería mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();

    } );

    test( 'debería ejecutarse el eventStartDeleteAction', () => {

        wrapper.find( 'button' ).simulate( 'click' );

        // wrapper.find('button').prop('onClick')(); // la otra manera de hacerlo

        expect( store.dispatch ).toHaveBeenCalled(); // No está mal pero queremos saber si se ha ejecutado eventStartDeleteAction

        // Para ello creamos un mock de eventStartDeleteAction

        expect( eventStartDeleteAction ).toHaveBeenCalled();

    } );

} );
