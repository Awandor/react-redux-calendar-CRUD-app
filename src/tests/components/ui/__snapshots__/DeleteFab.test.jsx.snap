// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <DeleteFab /> deber√≠a mostrarse correctamente 1`] = `
<button
  className="btn btn-danger fab-delete"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
</button>
`;
